hasktorch/src/Torch/Typed/Vision.hs

```diff
@@ -140,3 +162,372 @@ initMnist path = do
+        let (fptr, len) = V.unsafeToForeignPtr0 vec
+            whc = w * h * c * dtype_size
+        F.withForeignPtr fptr $ \ptr2 -> do
+          BSI.memcpy (F.castPtr ptr1) (F.castPtr ptr2) whc
+          print t
+          return $ toDevice @device @'(D.CPU, 0)
+                    . toDType @outDtype @dtype'
+                    . resize2d @outShape @"bilinear" True 
+                    . hwc2chw @'(D.CPU, 0) @dtype' @shape
+                    $ t 
+    createTensorU16to32 :: 
+      forall shape outShape outDtype dtype' device n' c' h' w' n0 c0 h0 w0 vec. 
+        (KnownShape shape, All KnownNat shape, 
+         KnownShape outShape, All KnownNat outShape,
+         KnownDType outDtype, KnownDType dtype', KnownDevice device, V.Storable vec,
+         (n' ': c' ': h' ': w' ': '[]) ~ outShape,
+         (n0 ': h0 ': w0 ': c0 ': '[]) ~ shape,
+         n' ~ n0,
+         c' ~ c0
+        ) =>  
+        V.Vector vec -> 
+        IO (Tensor device outDtype outShape)
+    createTensorU16to32 vec = do
+      let [_, h, w, c] = shapeVal @shape
+      t <- empty @shape @dtype' @'(D.CPU, 0)
+      withTensorPtr t $ \ptr1 -> do
+        let (fptr, len) = V.unsafeToForeignPtr0 vec
+            whc = fromIntegral $ w * h * c
+        F.withForeignPtr fptr $ \ptr2 -> do
+          let src = F.castPtr ptr2
+              dst = F.castPtr ptr1
+          [C.block| void {
+              uint16_t* src = $(uint16_t* src);
+              int32_t* dst = $(int32_t* dst);
+              for(int i=0;i<$(int whc);i++){
+                 dst[i] = src[i];
+              }
+          } |]
+          return $ toDevice @device @'(D.CPU, 0)
+                    . toDType @outDtype @dtype'
+                    . resize2d @outShape @"bilinear" True 
+                    . hwc2chw @'(D.CPU, 0) @dtype' @shape
+                    $ t 
+    createTensorU32to64 :: 
+      forall shape outShape outDtype dtype' device n' c' h' w' n0 c0 h0 w0 vec. 
+        (KnownShape shape, All KnownNat shape, 
+         KnownShape outShape, All KnownNat outShape,
+         KnownDType outDtype, KnownDType dtype', KnownDevice device, V.Storable vec,
+         (n' ': c' ': h' ': w' ': '[]) ~ outShape,
+         (n0 ': h0 ': w0 ': c0 ': '[]) ~ shape,
+         n' ~ n0,
+         c' ~ c0
+        ) =>
+        V.Vector vec -> 
+        IO (Tensor device outDtype outShape)
+    createTensorU32to64 vec = do
+      let [_, h, w, c] = shapeVal @shape
+      t <- empty @shape @dtype' @'(D.CPU, 0)
+      withTensorPtr t $ \ptr1 -> do
+        let (fptr, len) = V.unsafeToForeignPtr0 vec
+            whc = fromIntegral $ w * h * c
+        F.withForeignPtr fptr $ \ptr2 -> do
+          let src = F.castPtr ptr2
+              dst = F.castPtr ptr1
+          [C.block| void {
+              uint32_t* src = $(uint32_t* src);
+              int64_t* dst = $(int64_t* dst);
+              for(int i=0;i<$(int whc);i++){
+                 dst[i] = src[i];
+              }
+          } |]
```

Comment:
@junjihashimoto can you review this wrt to memory safety? I'm not sure this is garbage collected.

Reply:
although there are no allocations here, so probably it's fine

Reply:
In this case, `@dtype'` should be `D.Int64`, because the pointer of `dst` is `int64_t`'s array.

This is the same as `Torch.Vision`'s one.
If you can merge `createTensorU32to64` of both, try it.
https://github.com/hasktorch/hasktorch/blob/master/hasktorch/src/Torch/Vision.hs#L563-L578